src/TTextEdit.cpp

```diff
@@ -989,7 +989,7 @@ int TTextEdit::convertMouseXToBufferX(const int mouseX, const int lineNumber) co
     if (lineNumber < mpBuffer->lineBuffer.size()) {
         int characterWidth = 1;
         int currentX = 0;
-        for (const auto character : mpBuffer->lineBuffer.at(lineNumber)) {
+        for (const auto& character : mpBuffer->lineBuffer.at(lineNumber)) {
```

Comment:
:warning: Actually this is broken for a different reason - it is iterating through every `QChar` in the line of text and passing that to `(uint) TTextEdit::getGraphemeBaseCharacter(const QString& str) const` - which is wrong, *wrong*, **wrong**! That method's purpose is to return the first Unicode codepoint (the base character for the first grapheme) in the `QString` given to it.  When that method was designed it was expected that the `QString` being passed would be a single grapheme at a time (which may contain **more** that one Unicode code point) as extracted by using a `QTextBoundaryFinder` on the source text line.  

Passing every single `QChar` in a line to `getGraphemeWidth(...)` - which is what happens further down in - is likely to be what is spamming the debug output with `TTextEdit::Xxxxx(...) WARN - trying to get width of a Unicode character which is unprintable, codepoint number: U+XXXX` when it runs into a non-BMP codepoint (such as many modern Emojis) as that is represented by a pair of `QChars` (a *High* surrogate followed by a *Low* one in the UTF-16 encoding that is implicit in the `QString` class) - it will also barf at combining diacriticals as those should not be appearing as the first codepoint in a grapheme! :open_mouth:

Compare the usage here and in `(void) TTextEdit::slot_copySelectionToClipboardImage()` to the *correct* usage in `(int) TTextEdit::drawGrapheme(QPainter& painter, const QPoint& cursor, const QString& grapheme, int column, TChar& charStyle) const` and it's call (twice) from `(void) TTextEdit::drawLine(QPainter& painter, int lineNumber, int lineOfScreen) const` which DOE iterator through the lines in `(QStringList) TBuffer::lineBuffer` a grapheme at a time...

I'll take a closer look at these two suspect bits over the weekend and see if I can clean things up for Monday morning...

Reply:
Have a look at the screenshots - it clearly works for i18n :wink: It doesn't work for emoji's and other non-BMP codepoints because I'm not passing the right data in that usecase, but we can very clearly see that it works correctly for Chinese.

Let me know if you get the new PR up by Monday - Tuesday July 30th is the cutoff date for new PRs for 4.0, so if you're not able to make it by then we should merge this which is better than nothing for i18n. Not for emoji's but that is not my target here.